### PR TITLE
SDIT-739 Reconciliation endpoint

### PR DESCRIPTION
**Pull request by andymarke**

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -15,12 +15,14 @@ dependencies {
   implementation("org.springframework.boot:spring-boot-starter-security")
   implementation("org.springframework.boot:spring-boot-starter-oauth2-resource-server")
   implementation("org.springframework.boot:spring-boot-starter-oauth2-client")
+  implementation("org.springframework.data:spring-data-commons:3.0.5")
```
Comment (andymarke) [inline build.gradle.kts:18]: I just need the PageImpl interface

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/incentives/IncentivesReconciliationService.kt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            mapOf(
+              "page" to page.first.toString(),
+            ),
+          )
+          log.error("Unable to match entire page of prisoners: $page", it)
+        }.getOrElse { emptyList() }
+      log.info("Page requested: $page, with ${activePrisoners.size} active prisoners")
+
+      withContext(Dispatchers.Unconfined) {
+        activePrisoners.map { async { checkBookingIncentiveMatch(it) } }
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/incentives/IncentivesReconciliationService.kt:41]: run the page on 10 in parallel

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/incentives/IncentivesReconciliationService.kt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      }.awaitAll().filterNotNull()
+    }
+  }
+
+  private suspend fun checkBookingIncentiveMatch(prisonerId: ActivePrisonerId): MismatchIncentiveLevel? = runCatching {
+    log.info("Checking booking: ${prisonerId.bookingId}")
+    val (nomisIncentiveLevel, dpsIncentiveLevel) = withContext(Dispatchers.Unconfined) {
+      async { nomisApiService.getCurrentIncentive(prisonerId.bookingId) }.await() to
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/incentives/IncentivesReconciliationService.kt:49]: need to check this is running in parallel

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/incentives/IncentivesResource.kt [hunk 4]
```diff
@@ -1,18 +1,37 @@
...
+  }
+
+  @OptIn(DelicateCoroutinesApi::class)
   @PutMapping("/incentives/reports/reconciliation")
   @ResponseStatus(HttpStatus.ACCEPTED)
   fun generateIncentiveReconciliationReport() {
     telemetryClient.trackEvent("incentives-reports-reconciliation-requested")
-    telemetryClient.trackEvent("incentives-reports-reconciliation-report", mapOf("status" to "success", "issues-count" to "0"))
+    log.info("Incentives reconciliation report requested")
+    GlobalScope.launch {
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/incentives/IncentivesResource.kt:26]: run is coroutine - fire and forget

Comment (mikehalmamoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/incentives/IncentivesResource.kt:28]: Have you read this? https://elizarov.medium.com/the-reason-to-avoid-globalscope-835337445abc

**Review by mikehalmamoj — APPROVED** [review]

**Pull request merged**